### PR TITLE
fix(amm): buy_price calculation

### PR DESCRIPTION
**Pull request by Roznovjak**
## Description
Fee in buy_price calculation is wrongly assigned to the `amount_out`. The correct operation order is to compute `amount_in` first, and then apply fee to this value in the next step.

### CODE REVIEW

File: pallets/xyk/src/lib.rs [hunk 1]
```diff
@@ -604,21 +603,25 @@ impl<T: Config> AMM<T::AccountId, AssetId, AssetPair, Balance> for Pallet<T> {
...
 		let transfer_fee = if discount {
 			Self::calculate_discounted_fee(amount)?
 		} else {
 			Self::calculate_fee(amount)?
 		};
 
-		let sale_price = hydra_dx_math::calculate_out_given_in(asset_in_total, asset_out_total, amount - transfer_fee)
-			.map_err(|_| Error::<T>::SellAssetAmountInvalid)?;
+		let sale_price =
+			hydra_dx_math::calculate_out_given_in(asset_in_reserve, asset_out_reserve, amount - transfer_fee)
```
Comment (enthusiastmartin) [inline pallets/xyk/src/lib.rs:621]: Let's fix this as well and use checked math for fee subtraction

File: pallets/xyk/src/lib.rs [hunk 2]
```diff
@@ -604,21 +603,25 @@ impl<T: Config> AMM<T::AccountId, AssetId, AssetPair, Balance> for Pallet<T> {
...
+				.map_err(|_| Error::<T>::SellAssetAmountInvalid)?;
 
-		ensure!(asset_out_total >= sale_price, Error::<T>::InsufficientAssetBalance);
+		ensure!(asset_out_reserve > sale_price, Error::<T>::InsufficientAssetBalance);
```
Comment (enthusiastmartin) [inline pallets/xyk/src/lib.rs:624]: why not >=?

File: pallets/xyk/src/lib.rs [hunk 3]
```diff
@@ -731,27 +734,25 @@ impl<T: Config> AMM<T::AccountId, AssetId, AssetPair, Balance> for Pallet<T> {
...
+			.ok_or(Error::<T>::BuyAssetAmountInvalid)?;
 
 		ensure!(
-			T::Currency::free_balance(assets.asset_in, who) >= buy_price,
+			T::Currency::free_balance(assets.asset_in, who) >= buy_price_with_fee,
 			Error::<T>::InsufficientAssetBalance
 		);
 
-		ensure!(max_limit >= buy_price, Error::<T>::AssetBalanceLimitExceeded);
+		ensure!(max_limit >= buy_price_with_fee, Error::<T>::AssetBalanceLimitExceeded);
```
Comment (enthusiastmartin) [inline pallets/xyk/src/lib.rs:755]: Perhaps you can move this check before the free balance check.

**Review by enthusiastmartin — COMMENTED** [review]
Just few minor comments. Otherwise looks good to me.

**Comment by enthusiastmartin** [comment]
Btw i dont think this is a breaking change. IT is just implementation change but it does not break anything.

File: pallets/xyk/src/lib.rs (hunk 2 above)
Comment (Roznovjak) [reply pallets/xyk/src/lib.rs:624]: just to be sure that there is always some liquidity left in the pool

**Comment by Roznovjak** [comment]
You are right. A change in behavior to fix a bug is not considered as a breaking change.

File: pallets/xyk/src/lib.rs (hunk 1 above)
Comment (Roznovjak) [reply pallets/xyk/src/lib.rs]: done

File: pallets/xyk/src/lib.rs (hunk 3 above)
Comment (Roznovjak) [reply pallets/xyk/src/lib.rs]: done

**Review by enthusiastmartin — APPROVED** [review]

**Pull request merged**